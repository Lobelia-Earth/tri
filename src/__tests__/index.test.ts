import { tri } from '../index';

describe('Process', () => {
  // ====================================
  // Basic
  // ====================================
  it('null', () => {
    expect(tri(null)).toBeNull();
  });

  it('undefined', () => {
    expect(tri(undefined)).toBeUndefined();
  });

  it('tree with no definitions', () => {
    const tree = {
      a: {
        b: { c: 3, d: 'foo', e: null, f: true, g: undefined },
        h: [3, 4, 'a'],
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('removes $definitions', () => {
    const tree = { a: 3, $definitions: { foo: 4 } };
    expect(tri(tree)).toMatchSnapshot();
  });

  // ====================================
  // Interpolation
  // ====================================
  it('simple interpolation', () => {
    const tree = {
      a: { b: 'Length: <<$.len>> <<$.units>>' },
      $definitions: { len: 7, units: 'm' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('simple interpolation without escapes', () => {
    const tree = {
      a: { b: 'Comparisons <<$.lt>> <<$.gt>>' },
      $definitions: { lt: '<', gt: '>' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('nested interpolation', () => {
    const tree = {
      a: 'Hello <<$.foo>>',
      $definitions: {
        foo: '<<$.bar>>!',
        bar: 'world',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  // ====================================
  // Call shortcut
  // ====================================
  it('call shortcut', () => {
    const tree = {
      a: '$foo',
      $definitions: { foo: 'Some fixed text' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call shortcut, optional', () => {
    const tree = {
      a: '$foo?',
      $definitions: { foo: 'Some fixed text' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call shortcut with invalid name throws', () => {
    const tree = {
      a: '$fuzz',
      $definitions: { foo: 'Some fixed text' },
    };
    expect(() => tri(tree)).toThrowError();
  });

  it('call shortcut, optional, with invalid name does not throw', () => {
    const tree = { a: '$fuzz?' };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('nested call shortcuts', () => {
    const tree = {
      a: '$foo',
      $definitions: { foo: '$bar', bar: 'Final text' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('exploded short call', () => {
    const tree = {
      a: ['$*foo', '$*bar'],
      $definitions: {
        foo: ['$*qux', 3],
        bar: [4, '$theEnd'],
        qux: [1, 2],
        theEnd: 'finished!',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('exploded short call returning non-array throws', () => {
    const tree = {
      a: ['$*foo'],
      $definitions: { foo: 'Some fixed text' },
    };
    expect(() => tri(tree)).toThrowError();
  });

  // ====================================
  // Call shortcut + interpolation
  // ====================================
  it('nested shortcut and interpolation', () => {
    const tree = {
      a: '$foo',
      $definitions: { foo: '$bar', bar: 'Hello <<$.bass>>', bass: 'world!' },
      // NOTE: it doesn't work the other way around (shortcut inside interpolation)
      // Workaround: just use interpolation inside interpolation in this case
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('nested interpolation and shortcut', () => {
    const tree = {
      a: '$foo',
      $definitions: {
        foo: '<<$.prefix>> <<$.bar>>',
        prefix: 'Hello',
        bar: '$bass',
        bass: 'world!',
      },
    };
    expect(tri(tree).a).toEqual('Hello world!');
  });

  // ====================================
  // Calls
  // ====================================
  it('call', () => {
    const tree = {
      a: { $call: 'foo' },
      $definitions: { foo: 3 },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call, optional', () => {
    const tree = {
      a: { $call: 'foo', $optional: true },
      $definitions: { foo: 3 },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call with invalid name throws', () => {
    const tree = { a: { $call: 'foo' } };
    expect(() => tri(tree)).toThrowError();
  });

  it('call, optional, with invalid name does not throw', () => {
    const tree = {
      a: { $call: 'fuzz', $optional: true },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call with params', () => {
    const tree = {
      a: { $call: 'fn', $params: { param1: 7, param2: { foo: true } } },
      $definitions: {
        fn: {
          result1: '$param1',
          result2: '$param2',
          result3: 'My age is <<$.param1>>',
        },
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('nested calls with params', () => {
    const tree = {
      a: { $call: 'fn1', $params: { param1: 7 } },
      $definitions: {
        fn1: { $call: 'fn2', $params: { param2: 10 } },
        fn2: { result1: '$param1', result2: '$param2' },
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('exploded call', () => {
    const tree = {
      a: [
        { $call: 'foo', $explode: true },
        { $call: 'bar', $explode: true },
      ],
      $definitions: {
        foo: ['$*qux', 3],
        bar: [4, '$theEnd'],
        qux: [1, 2],
        theEnd: 'finished!',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('exploded call returning non-array throws', () => {
    const tree = {
      a: [{ $call: 'foo', $explode: true }],
      $definitions: { foo: 'Some fixed text' },
    };
    expect(() => tri(tree)).toThrowError();
  });

  it('call with $forEach', () => {
    const tree = {
      a: {
        $call: 'foo',
        $forEach: [
          { name: 'Pau' },
          { name: 'Théo' },
          { name: 'Sascha' },
          { name: 'Jordi' },
          { name: 'Aleix' },
          { name: 'Guille' },
        ],
      },
      $definitions: {
        salutation: 'Hello',
        foo: '<<$.salutation>>, <<$.name>>!',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call with $forEach and $params', () => {
    const tree = {
      a: {
        $call: 'foo',
        $forEach: [{ name: 'Pau' }, { name: 'Théo' }],
        $params: { salutation: 'Hi' },
      },
      $definitions: {
        foo: '<<$.salutation>>, <<$.name>>!',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call with $forEach but no array throws', () => {
    const tree = {
      a: {
        $call: 'foo',
        $forEach: { name: 'Pau' }, // should be an array!
      },
      $definitions: { foo: 'Hi, <<$.name>>!' },
    };
    expect(() => tri(tree)).toThrowError();
  });

  it('call with $forEach and $explode: true', () => {
    const tree = {
      a: [
        {
          $call: 'foo',
          $explode: true,
          $forEach: [{ item: 'A' }, { item: 'B' }, { item: 'C' }],
        },
      ],
      $definitions: { foo: ['<<$.item>>1', '<<$.item>>2'] },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('call with $forEach and $explode: true, but call returns an object', () => {
    const tree = {
      a: [
        {
          $call: 'foo',
          $explode: true,
          $forEach: [{ item: 'A' }, { item: 'B' }, { item: 'C' }],
        },
      ],
      $definitions: { foo: '<<$.item>>1' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  // ====================================
  // Merge
  // ====================================
  it('$merge without calls', () => {
    const tree = { a: { $merge: [{ b: 3 }, { c: 4 }] } };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$merge with calls', () => {
    const tree = {
      a: {
        $merge: [
          '$fn1',
          { $call: 'fn2', $params: { param: 'foo' } },
          { $call: 'fn3', $params: { param: 3.14 } },
        ],
      },
      $definitions: {
        fn1: { a: 3 },
        fn2: { b: 'Hi <<$.param>>!' },
        fn3: { c: '$param' },
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$merge throws without array', () => {
    const tree = { a: { $merge: { b: 3 } } };
    expect(() => tri(tree)).toThrow();
  });

  // ====================================
  // Concatenate
  // ====================================
  it('$concatenate without calls', () => {
    const tree = { a: { $concatenate: [1, 2, [3, 4], 5, [[6, 7]]] } };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$concatenate with calls', () => {
    const tree = {
      a: {
        $concatenate: [
          '$salute',
          { $call: 'say', $params: { param: 'Nice to see you' } },
          { $call: 'ask', $params: { param: 'What time is it' } },
        ],
      },
      $definitions: {
        salute: ['Hi', '----'],
        say: ['<<$.param>>!', '----'],
        ask: ['<<$.param>>?', '----'],
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$concatenate throws without array', () => {
    const tree = { a: { $concatenate: { b: 3 } } };
    expect(() => tri(tree)).toThrow();
  });

  // ====================================
  // CSV
  // ====================================
  it('$csv', () => {
    const tree = {
      a: {
        $csv: `a,b,c
1,a,true
,b,false`,
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$csv with relative path', () => {
    const tree = {
      a: {
        $csv: './src/__tests__/data.csv',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$csv with absolute path', () => {
    const tree = {
      a: {
        $csv: `file://${process.cwd()}/src/__tests__/data.csv`,
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$csv with missing columns throws', () => {
    const tree = {
      a: {
        $csv: `a,b,c
1,a,2
c,b`,
      },
    };
    expect(() => tri(tree)).toThrow();
  });

  it('$csv with extra columns throws', () => {
    const tree = {
      a: {
        $csv: `a,b,c
1,a,3
2,b,2,f`,
      },
    };
    expect(() => tri(tree)).toThrow();
  });

  it('$csv without content throws', () => {
    const tree = { a: { $csv: null } };
    expect(() => tri(tree)).toThrow();
  });

  it('$csv with $json parsing', () => {
    const tree = {
      a: {
        $csv: `foo,jsonField
a,"{""foo"":true}"
b,`,
        $json: ['jsonField'],
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$csv with interpolated content', () => {
    const tree = {
      a: {
        $csv: `a,b,c
1,a,<<$.myVar>>`,
      },
      $definitions: { myVar: 'true' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$csv with $forEach', () => {
    const tree = {
      a: {
        $call: 'myFunction',
        $forEach: { $csv: 'name,age\n"John",30\n"Jane",28' },
      },
      $definitions: { myFunction: { a: '$name', b: '$age' } },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  // ====================================
  // YAML
  // ====================================
  it('$yaml with relative path', () => {
    const tree = {
      a: {
        $yaml: './src/__tests__/data.yaml',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$yaml with absolute path', () => {
    const tree = {
      a: {
        $yaml: `file://${process.cwd()}/src/__tests__/data.yaml`,
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$yaml without content throws', () => {
    const tree = { a: { $yaml: null } };
    expect(() => tri(tree)).toThrow();
  });

  it('$yaml calling parent definitions', () => {
    const tree = {
      a: {
        $yaml: `foo: $foo`,
      },
      $definitions: { foo: 'bar' },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  it('$yaml with array', () => {
    const tree = {
      a: {
        $yaml: '- foo\n- bar\n- baz',
      },
    };
    expect(tri(tree)).toMatchSnapshot();
  });

  // ====================================
  // Deep copies
  // ====================================
  it('deep copies may exist by default', () => {
    const tree = {
      a: '$foo',
      b: '$foo',
      $definitions: { foo: { bar: 3 } },
    };
    const res = tri(tree);
    expect(res.a === res.b).toBeTruthy();
  });

  it('deep copies can be disabled', () => {
    const tree = {
      a: '$foo',
      b: '$foo',
      $definitions: { foo: { bar: 3 } },
    };
    const res = tri(tree, { noDeepCopies: true });
    expect(res.a === res.b).toBeFalsy();
  });
});
