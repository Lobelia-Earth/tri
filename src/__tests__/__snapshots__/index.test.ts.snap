// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process $concatenate with calls 1`] = `
Object {
  "a": Array [
    "Hi",
    "----",
    "Nice to see you!",
    "----",
    "What time is it?",
    "----",
  ],
}
`;

exports[`Process $concatenate without calls 1`] = `
Object {
  "a": Array [
    1,
    2,
    3,
    4,
    5,
    Array [
      6,
      7,
    ],
  ],
}
`;

exports[`Process $csv 1`] = `
Object {
  "a": Array [
    Object {
      "a": 1,
      "b": "a",
      "c": true,
    },
    Object {
      "a": null,
      "b": "b",
      "c": false,
    },
  ],
}
`;

exports[`Process $csv with $forEach 1`] = `
Object {
  "a": Array [
    Object {
      "a": "John",
      "b": 30,
    },
    Object {
      "a": "Jane",
      "b": 28,
    },
  ],
}
`;

exports[`Process $csv with $json parsing 1`] = `
Object {
  "a": Array [
    Object {
      "foo": "a",
      "jsonField": Object {
        "foo": true,
      },
    },
    Object {
      "foo": "b",
      "jsonField": null,
    },
  ],
}
`;

exports[`Process $csv with absolute path 1`] = `
Object {
  "a": Array [
    Object {
      "a": 1,
      "b": "a",
      "c": true,
    },
    Object {
      "a": null,
      "b": "b",
      "c": false,
    },
  ],
}
`;

exports[`Process $csv with interpolated content 1`] = `
Object {
  "a": Array [
    Object {
      "a": 1,
      "b": "a",
      "c": true,
    },
  ],
}
`;

exports[`Process $csv with relative path 1`] = `
Object {
  "a": Array [
    Object {
      "a": 1,
      "b": "a",
      "c": true,
    },
    Object {
      "a": null,
      "b": "b",
      "c": false,
    },
  ],
}
`;

exports[`Process $merge with calls 1`] = `
Object {
  "a": Object {
    "a": 3,
    "b": "Hi foo!",
    "c": 3.14,
  },
}
`;

exports[`Process $merge without calls 1`] = `
Object {
  "a": Object {
    "b": 3,
    "c": 4,
  },
}
`;

exports[`Process $yaml calling parent definitions 1`] = `
Object {
  "a": Object {
    "foo": "bar",
  },
}
`;

exports[`Process $yaml with absolute path 1`] = `
Object {
  "a": Object {
    "a": 3,
    "b": Object {
      "c": "foo",
      "d": 4,
    },
  },
}
`;

exports[`Process $yaml with array 1`] = `
Object {
  "a": Array [
    "foo",
    "bar",
    "baz",
  ],
}
`;

exports[`Process $yaml with relative path 1`] = `
Object {
  "a": Object {
    "a": 3,
    "b": Object {
      "c": "foo",
      "d": 4,
    },
  },
}
`;

exports[`Process call 1`] = `
Object {
  "a": 3,
}
`;

exports[`Process call shortcut 1`] = `
Object {
  "a": "Some fixed text",
}
`;

exports[`Process call shortcut, optional 1`] = `
Object {
  "a": "Some fixed text",
}
`;

exports[`Process call shortcut, optional, with invalid name does not throw 1`] = `
Object {
  "a": undefined,
}
`;

exports[`Process call with $forEach 1`] = `
Object {
  "a": Array [
    "Hello, Pau!",
    "Hello, Théo!",
    "Hello, Sascha!",
    "Hello, Jordi!",
    "Hello, Aleix!",
    "Hello, Guille!",
  ],
}
`;

exports[`Process call with $forEach and $explode: true 1`] = `
Object {
  "a": Array [
    "A1",
    "A2",
    "B1",
    "B2",
    "C1",
    "C2",
  ],
}
`;

exports[`Process call with $forEach and $explode: true, but call returns an object 1`] = `
Object {
  "a": Array [
    "A1",
    "B1",
    "C1",
  ],
}
`;

exports[`Process call with $forEach and $params 1`] = `
Object {
  "a": Array [
    "Hi, Pau!",
    "Hi, Théo!",
  ],
}
`;

exports[`Process call with params 1`] = `
Object {
  "a": Object {
    "result1": 7,
    "result2": Object {
      "foo": true,
    },
    "result3": "My age is 7",
  },
}
`;

exports[`Process call, optional 1`] = `
Object {
  "a": 3,
}
`;

exports[`Process call, optional, with invalid name does not throw 1`] = `
Object {
  "a": undefined,
}
`;

exports[`Process exploded call 1`] = `
Object {
  "a": Array [
    1,
    2,
    3,
    4,
    "finished!",
  ],
}
`;

exports[`Process exploded short call 1`] = `
Object {
  "a": Array [
    1,
    2,
    3,
    4,
    "finished!",
  ],
}
`;

exports[`Process nested call shortcuts 1`] = `
Object {
  "a": "Final text",
}
`;

exports[`Process nested calls with params 1`] = `
Object {
  "a": Object {
    "result1": 7,
    "result2": 10,
  },
}
`;

exports[`Process nested interpolation 1`] = `
Object {
  "a": "Hello world!",
}
`;

exports[`Process nested shortcut and interpolation 1`] = `
Object {
  "a": "Hello world!",
}
`;

exports[`Process removes $definitions 1`] = `
Object {
  "a": 3,
}
`;

exports[`Process simple interpolation 1`] = `
Object {
  "a": Object {
    "b": "Length: 7 m",
  },
}
`;

exports[`Process simple interpolation without escapes 1`] = `
Object {
  "a": Object {
    "b": "Comparisons < >",
  },
}
`;

exports[`Process tree with no definitions 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": 3,
      "d": "foo",
      "e": null,
      "f": true,
      "g": undefined,
    },
    "h": Array [
      3,
      4,
      "a",
    ],
  },
}
`;
